Panic during yield inside over+switch

## tl;dr

I bumped into a panic when attempting to `yield` a value from inside a `switch` that's inside an `over`. If I remove the `switch` layer the panic no longer triggers.

## Details

Repro is with Zed commit 9f61b4e and test data files [results1.csv](https://github.com/brimdata/zed/files/14273912/results1.csv) and [results2.csv](https://github.com/brimdata/zed/files/14273914/results2.csv).

Running the contents of `repro.zed`

```
file results1.csv format csv
| inner join (
  file results2.csv format csv
) on machine_name=machine_name zed_results:=this
| over this with zed_results => (
  switch (
    case true => yield zed_results
  )
)
```

triggers a panic:

```
$ zq -version
Version: v1.13.0-12-g9f61b4ea

$ zq -I repro.zed 
panic: runtime error: index out of range [0] with length 0
goroutine 1 [running]:
runtime/debug.Stack()
	/usr/local/opt/go/libexec/src/runtime/debug/stack.go:24 +0x5e
github.com/brimdata/zed/runtime/sam/op.(*Catcher).Pull.func1()
	/Users/phil/work/zed/runtime/sam/op/catcher.go:25 +0x3d
panic({0x254d340?, 0xc000a8e000?})
	/usr/local/opt/go/libexec/src/runtime/panic.go:914 +0x21f
github.com/brimdata/zed/runtime/sam/expr.Var.Eval(...)
	/Users/phil/work/zed/runtime/sam/expr/var.go:14
github.com/brimdata/zed/runtime/sam/op/yield.(*Op).Pull(0xc0005fea50, 0x10?)
	/Users/phil/work/zed/runtime/sam/op/yield/yield.go:31 +0x432
github.com/brimdata/zed/runtime/sam/op/traverse.(*Exit).pullPlatoon(0xc0005fea80)
	/Users/phil/work/zed/runtime/sam/op/traverse/scope.go:222 +0x32
github.com/brimdata/zed/runtime/sam/op/traverse.(*Exit).Pull(0xc0005fea80, 0x5?)
	/Users/phil/work/zed/runtime/sam/op/traverse/scope.go:208 +0x2b
github.com/brimdata/zed/runtime/sam/op.(*Single).Pull(0xc0005e0f60, 0x0)
	/Users/phil/work/zed/runtime/sam/op/mux.go:134 +0x2b
github.com/brimdata/zed/runtime/sam/op.(*Catcher).Pull(0x3cbc108?, 0x0?)
	/Users/phil/work/zed/runtime/sam/op/catcher.go:28 +0x62
github.com/brimdata/zed/runtime/exec.(*Query).Pull(0x4ab66b58?, 0xd8?)
	/Users/phil/work/zed/runtime/exec/query.go:59 +0x3e
github.com/brimdata/zed/zbuf.CopyPuller({0x4ab66b58, 0xc0005e0a98}, {0x28be060, 0xc0005feab0})
	/Users/phil/work/zed/zbuf/batch.go:162 +0x4d
github.com/brimdata/zed/cli/zq.(*Command).Run(0xc0004caa80, {0xc000040740, 0x0, 0x0})
	/Users/phil/work/zed/cli/zq/command.go:162 +0x79e
github.com/brimdata/zed/pkg/charm.path.run({0xc0006ae300?, 0x1, 0x1}, {0xc000040740?, 0x0, 0x0?})
	/Users/phil/work/zed/pkg/charm/path.go:11 +0x79
github.com/brimdata/zed/pkg/charm.(*Spec).ExecRoot(0x100874b?, {0xc000040730, 0x2, 0x2})
	/Users/phil/work/zed/pkg/charm/charm.go:63 +0x3f
main.main()
	/Users/phil/work/zed/cmd/zq/main.go:11 +0x5b
```

If I drop the `switch` layer such as in `ok.zed`

```
file results1.csv format csv
| inner join (
  file results2.csv format csv
) on machine_name=machine_name zed_results:=this
| over this with zed_results => (
  yield zed_results
)
```

it does not panic:

```
$ zq -I ok.zed 
{machine_name:"anansi",cpu_min:0.,cpu_max:22.5,cpu_avg:0.14268608779353878,net_in_min:938.9500122070312,net_in_max:2.35533575e+06,net_in_avg:9177.160319882514,net_out_min:216.8800048828125,net_out_max:2.35453625e+06,net_out_avg:13467.316303191717}
{machine_name:"anansi",cpu_min:0.,cpu_max:22.5,cpu_avg:0.14268608779353878,net_in_min:938.9500122070312,net_in_max:2.35533575e+06,net_in_avg:9177.160319882514,net_out_min:216.8800048828125,net_out_max:2.35453625e+06,net_out_avg:13467.316303191717}
...
```

--- a/runtime/sam/op/switcher/switch.go
+++ b/runtime/sam/op/switcher/switch.go
@@ -68,7 +68,7 @@ func (s *Selector) Forward(router *op.Router, batch zbuf.Batch) bool {
 			// outgoing batch so we don't send these slices
 			// through GC.
 			batch.Ref()
-			out := zbuf.NewArray(c.vals)
+			out := zbuf.NewBatch(batch, c.vals)
 			c.vals = nil
 			if ok := router.Send(c.route, out, nil); !ok {
 				return false

--- a/runtime/sam/op/switcher/ztests/switch-vars.yaml
+++ b/runtime/sam/op/switcher/ztests/switch-vars.yaml
@@ -1,0 +1,12 @@
+zed: |
+  over this with foo = {foo:"bar"} => (
+    switch (
+      case x % 2 == 0 => yield {x, ...foo}
+    )
+  )
+
+input: "[{x:1},{x:2},{x:3},{x:4}]"
+
+output: |
+  {x:2,foo:"bar"}
+  {x:4,foo:"bar"}

